centralise rose cylc docs (#1033)
ACCESS-Hive/access-hive.github.io: docs/models/run-a-model/workflow-rose-cylc.md

# The Rose/Cylc Workflow Management Tool

> The _Rose/Cylc_ workflow management tool consists of two components- the [_Cylc_](https://niwa.co.nz/environmental-information/cylc-suite-engine) task engine, developed by the New Zealand National Institute of Water and Atmospheric Research (NIWA), and the [_Rose_](https://www.metoffice.gov.uk/research/approach/modelling-systems/rose) framework developed by the UK Met Office (UKMO) which configures tasks for the _Cylc_ engine. A workflow comprised of set of tasks configured by _Rose_ to run with the _Cylc_ engine is called a "suite".

Reviewer: ```suggestion
The _Rose/Cylc_ workflow management tool consists of two components: the [_Cylc_](https://cylc.github.io/) task engine, developed by the New Zealand National Institute of Water and Atmospheric Research (NIWA), and the [_Rose_](https://metomi.github.io/rose/doc/html/index.html) framework developed by the UK Met Office (UKMO) which configures tasks for the _Cylc_ engine.
```

I would use directly the links to Rose and Cylc documentations.

Also, the part:
> A workflow comprised of set of tasks configured by _Rose_ to run with the _Cylc_ engine is called a "suite".

is problematic because in Cylc8 "suite" becomes "workflow". Therefore, I would only use it when it's needed in specific paragraphs (not general as this one), so we might easily modify them when we write the same page for Cylc8.

Author: How set are you on the links being to the documentation, rather than the organisation page? This paragraph is really just a "What is it?" description, we don't need (or even want) users to go to the actual documentation yet. The organisation pages have links to the docs, so if users really want to read the docs, they can still go via the org page.

I think it's important to introduce the idea of a suite/workflow early, since we often refer to them. None of our models run on Cylc8 yet right? I propose keeping the wording on "suites" the same here, but add a warning along the lines of:

In Cylc8, the term "suite" is replaced by "workflow". We are in the process of transitioning to Cylc8. During this process, "suite" and "workflow" can be used interchangeably.

## Prerequisites

Before running a _Rose/Cylc_ suite on Gadi, you must:

1. [Set up your NCI account](/getting_started/set_up_nci_account).
2. Apply for a MOSRS account by getting in contact with [your local institutional sponsor](https://opus.nci.org.au/display/DAE/Prerequisites).
3. Join the following NCI projects:
    - [access](https://my.nci.org.au/mancini/project/access/join)
    - [hr22](https://my.nci.org.au/mancini/project/hr22/join)
    - [ki32](https://my.nci.org.au/mancini/project/ki32/join)
    - [ki32_mosrs](https://my.nci.org.au/mancini/project/ki32_mosrs/join)
    
    !!! tip
        To request membership for the _ki32_mosrs_ subproject, you need to:
        
        - already be member of the _ki32_ project
        {: style="list-style-type: disc"}
        - have a [MOSRS account](#mosrs-account)

>         {: style="list-style-type: disc"}

Reviewer: The format is good, but it needs to be made specific to this page. 
For MOSRS, I would change the phrasing to:
> The [Met Office Science Repository Service (MOSRS)](https://code.metoffice.gov.uk) is a server run by the UK Met Office (UKMO) to support collaborative development with other partners organisations. MOSRS contains the source code of some ACCESS models configurations and model components.<br>
    To apply for a _MOSRS_ account, please contact your [local institutional sponsor](https://opus.nci.org.au/display/DAE/Prerequisites).

 
## Launching a Cylc suite

There are two options for launching a _Rose/Cylc_ suite- either via the terminal directly, or via a Gadi [ARE VDI session](https://opus.nci.org.au/spaces/Help/pages/163250532/2.1.+Connecting+to+the+VDI). Both methods eventually end up in a terminal, so if you wish to use the terminal directory on a Gadi login node, skip directly to the [Set up a persistent session](#set-up-a-persistent-session) section.

!!! warning "X11 Forwarding"
    If you are going to connect via the terminal directly using SSH, it is recommended to connect with `ssh -X` to enable X11 forwarding. This allows the Rose and Cylc GUIs to be launched on your local machine.

### Launch ARE VDI Desktop

> The following options are recommended for your ARE VDI desktop session. Note that the session is not running the tasks within the suite, only launching the _Rose/Cylc_ tool, which will then dispatch the tasks within the suite to the compute queue.

Reviewer: ```suggestion
!!! tip
    The ARE VDI session does not run model tasks directly; it only runs _Rose_ and _Cylc_ tools. The model tasks are dispatched by _Cylc_ to the computing nodes. As a result, the ARE VDI session requires minimal CPU and memory resources.
```

Author: I think the first sentence "The following options are recommended for your ARE VDI desktop session." should remain outside the tip. The page should make sense without the tips and warnings, so I think it's still necessary to specify what these specifications are

Reviewer: I agree the page should make sense without admonition, and this `tip` doesn't list the specifications. It only gives reason about some of the specifications listed below. 
Users would still be able to run ARE using the specifications below with no issue even without reading this `tip`.

- **Walltime (hours)** &rarr; `2`<br>

>     Amount of hours the ARE VDI session will remain active for. This is only setup time, and does not reflect how long the actual suite will take to run.

Reviewer: ```suggestion
    Amount of hours the ARE VDI session will remain active for. This is only setup time, and does not reflect how long the actual experiment will take to run.
```

Author: Changed this to "configuration".

- **Queue** &rarr; `normalbw`
    
- **Compute Size** &rarr; `tiny` (1 CPU)<br>

- **Project** &rarr; a project of which you are a member.<br>
    The project must have allocated [_Service Units (SU)_](https://opus.nci.org.au/spaces/Help/pages/236881132/Allocations...) to run your simulation. Usually, but not always, this corresponds to your `${PROJECT}`.

- **Storage** &rarr; `gdata/access+gdata/xp65+gdata/hr22+gdata/ki32` (minimum)<br>

>     Storage locations on Gadi that will be visible during execution of the suite. The packages listed above are the minimum required to run _Rose/Cylc_. Different models will require different storage locations, typically to have access to input data. The additional locations will be listed under `storage` on the respective [Run a Model](https://docs.access-hive.org.au/models/run-a-model/) pages.

Reviewer: ```suggestion
    The storage folders listed above are the minimum required to run _Rose/Cylc_.
```
This only needs storage folders needed for the setup, not for the actual modelling (those are set within the suites/workflows).

Once the ARE VDI session opens in your browser, open the terminal by clicking the terminal icon at the top of the window. Use the terminal for the remainder of the instructions in this guide.

## Set up a persistent session

> NCI provides a service called [_persistent sessions_](https://opus.nci.org.au/spaces/Help/pages/241926895/Persistent+Sessions) to facilitate long running processes, like the _Rose/Cylc_ manager. Begin a new persistent session with

Reviewer: ```suggestion
NCI provides a service called [_persistent sessions_](https://opus.nci.org.au/spaces/Help/pages/241926895/Persistent+Sessions) to enable long-running processes. 

### Start a new persistent session {: .no-toc }
Start a new persistent session by running:
```

Author: I think it's worth making clear that Rose/Cylc is the long running process we want to enable

```
persistent-sessions start -p <project> <name>
```

> where `<project>` is the project you want to start the session under, and `<name>` is the name you want to give your persistent session. The persistent session is given a unique identifier (referred to here as `<persistent-session-uuid>`), which you will need to end the session later.

Reviewer: ```suggestion
where `<project>` is the project you want to start the session under, and `<name>` is the name you want to give your persistent session. 

!!! tip
    By omitting `-p <project>`, your default project `$PROJECT` will be used.<br>
    The project needs to have allocated _Service Units (SU)_ and it can be different from the project used to actually run the model experiment.

The newly created persistent session is assigned a unique identifier, referred to here as <persistent-session-uuid>.
```

Reviewer: I'm not 100% sure. 
And in fact I tested starting one with a project that doesn't have SU (`al33`) and it works without problems.

Yes I would remove that part. Thank you!

<terminal-window data="input">
    <terminal-line>persistent-sessions -p &lt;project&gt; start &lt;name&gt;</terminal-line>
    <terminal-line data="output">session &lt;persistent-session-uuid&gt; running - connect using</terminal-line>
    <terminal-line data="output">&emsp;ssh &lt;name&gt;.&lt;$USER&gt;.&lt;project&gt;.ps.gadi.nci.org.au</terminal-line>
</terminal-window>

> Once the session is started, it needs to be assigned to the model run. This is done by inserting the persistent session label into `~/.persistent-sessions/cylc-session`, which can be done manually or with the command

Reviewer: ```suggestion
Once the session is running, it needs to be assigned to _Cylc_. This is done by inserting the persistent session label into `~/.persistent-sessions/cylc-session`, which can be done by running the following command (after substituting `<name>` and `<project>` with the actual name and project used to create the persistent session):
```

```
cat > ~/.persistent-sessions/cylc-session <<< "<name>.${USER}.<project>.ps.gadi.nci.org.au"
```

You can check that this worked with

```
cat ~/.persistent-sessions/cylc-session
```

For example, if user `abc123` started a persistent session named `CM2` under the project `tm70`, then the command would be:

<terminal-window data="input">
    <terminal-line>cat > ~/.persistent-sessions/cylc-session <<< CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>
    <terminal-line data="input" linedelay="1000">cat ~/.persistent-sessions/cylc-session</terminal-line>
    <terminal-line data="output">CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>
</terminal-window>

Now the persistent session is set up, you can set up the _Rose/Cylc_ suite.

To list currently active suites, use:

```
persistent-sessions list
```

To end a specific session, use:

```
persistent-sessions kill <persistent-session-uuid>

> ```

Reviewer: ````suggestion
### Terminate a persistent sessions {: .no-toc }
To end a specific persistent session, use:

```
persistent-sessions kill <persistent-session-uuid>
```
````

## Rose/Cylc Setup

First, load the _Cylc_ module with:

```
module use /g/data/hr22/modulefiles
module load cylc7
```

> 

Reviewer: ```suggestion
### MOSRS authentication
```
I am not sure if this step will still be required when the GitHub migration is completed (I believe so at least to run models using the UM). 
Therefore, I would place it as a separate step before any other further step.

Please move here the section below about MOSRS authentication.

Next, retrieve the model configuration you want to run from a remote host. Depending on the specific model configuration, the host is _Github_ or _MOSRS_. The [Run a Model](https://docs.access-hive.org.au/models/run-a-model/) documentation for the specific model will specify which source to use for that configuration. Regardless of where the configuration comes from, it is recommended to store them in `~/roses/` (this happens automatically for suites pulled from MOSRS).

> ### Github Suites

Reviewer: ```suggestion
### Model configuration stored on GitHub
```

If the model configuration is hosted on Github, the Run a Model documentation will specify `repository` and `branch`, which describes where the default configuration is on Github. Move to the `~/roses` directory and pull the configuration from Github with:

```
git clone <repository> -b <branch>

> ```

Reviewer: ````suggestion
Get a copy of a model configuration hosted on the branch `<branch>`  of the Github repository `<repository>` by cloning the repository with:

```
git clone <repository> -b <branch>
```

For example, to get the ... configuration for ACCESS-...., you could run:

```
git clone .... -b ....
```
````
Do we have an example for this?

Reviewer: Hmm, yes unfortunately I think we cannot be too sure right now of how things will work after the migration. For this reason, for now I would keep this page referring to anything actual and already reproducible.

For the moment, I would keep this whole "GitHub" part commented (so it's easier to add when the time comes), and consider the configurations only to be stored on MOSRS. 
Whenever documentation on how to run AM3 is produced (or after we know what's the actual reproducible workflow), we can add this part back.

Author: You can pull the AM3 configuration and run it, it's just not officially even alpha released yet. I think this is a bit of a chicken-and-egg situation- we don't want to add these docs, with specific instructions on where to find the suite ID/github repo, without having an example to point to, but we don't want to change the other Run a model docs until this doc is online. I think we should keep the text here uncommented for both options. I think the current form is quite explicit.

The suite will appear in your current directory in the folder `<repository>`.

### MOSRS Suites

> If the model configuration is hosted on MORSRS, the Run a Model documentation will specify `suite-id` (potentially multiple suites), which describe the suite ID(s) on the MOSRS repository. To pull from MOSRS, you must first authenticate your MOSRS credentials with:

Author: Have adjusted the wording here, making clear that the respective model docs will specify the `suite-id` and `branch`.

```
mosrs-auth
```
<terminal-window>
    <terminal-line data="input">mosrs-auth</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: You need to enter your MOSRS credentials here so that GPG can cache your password.</terminal-line>
    <terminal-line>Please enter the MOSRS password for &lt;MOSRS-username&gt;:</terminal-line>
    <terminal-line lineDelay=1500>Checking your credentials using Subversion. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed Subversion with your credentials.</terminal-line>
    <terminal-line lineDelay=100><span style="color: #559cd5;">INFO</span>: Checking your credentials using rosie. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed rosie with your credentials.</terminal-line>
</terminal-window>

> Which will request your MOSRS username and password that you received when you created your MOSRS account. Once authenticated, pull the configuration from MOSRS with:

Reviewer: ```suggestion
!!! tip
    This command will request the MOSRS username and password that you received when you created your MOSRS account.
```

Reviewer: It can be laid out as a normal text too.

```
rosie checkout <suite-id>
```

> The suite will appear in your current directory in the folder `<suite-id>`.

Reviewer: ```suggestion
The configuration will be place in the `~/roses/<suite-id>/` folder.
```

Reviewer: I think the "GitHub" sentence should be removed because the behaviour is the `git clone` behaviour (i.e. create a directory in the current working directory, unless specific flags are specified).

Here I would keep it because the configuration ends up always in a that directory (not even sure if there's a flag to specify the destination path) even if your current directory is different, so I would make it clear.

> 

Reviewer: I think we should also mention the difference between `rosie checkout` and `rosie copy`

Author: I don't think this information should go here- I don't think new users should ever `rosie copy`?

Reviewer: Well not necessarily, a new user might still want to perturb an experiment and create a different one, so they would need to `rosie copy` to be able to re-upload the new experiment to MOSRS.

I think in general it's still a useful information that pertains to _Rose_/_Rosie_ so I would include it.

Author: How do you actually use `rosie copy`? The instructions on [rAM3](https://docs.access-hive.org.au/models/run-a-model/run-access-ram/#remote-copy) don't actually explain why you would want to `rosie copy`, or how to push changes to your local copy back to the remote.

Reviewer: `rosie copy` is to be used if you want to create your own suite based off an existing one. It will create a new remote suite (new ID) and clone the local copy.
To push changes back to the remote, you can use `fcm commit` (or `svn commit`). The `commit` command can be also shortened to `ci` (e.g., `fcm ci`).

I think this information is useful and should be added in this documentation.

Reviewer: I added this section in my updated review.

## Running the Suite

> Now that you've pulled the suite, you can start the suite by changing into the suite directory and running:

Author: Yes, this works for Github configurations. Once it's been cloned, I don't think rose sees the difference. I think it's better to say "run the configuration" instead of "run the model", since the model itself is only one task in the configuration

```
rose suite-run
```

> This launches the _Rose/Cylc_ suite. The _Cylc_ GUI should open which you can use to inspect the state of the suite (if you are running in the terminal directly and it doesn't appear, make sure you connected with [X11 forwarding](#x11-forwarding)). If you closed the GUI but want to re-open it, navigate to the suite directory and run:

Reviewer: ```suggestion
This will open the _Cylc_ GUI, which you can use to inspect the state of the experiment. 

!!! warning
    If you are running the model from a _Gadi_ login node and the _Cylc_ GUI doesn't open, make sure you connected using [X11 forwarding](#x11-forwarding)). 

If you closed the GUI and want to re-open it, navigate to the configuration directory and run:
```

Reviewer: Yes you're right, then I would add:

> This will start the configuration run and open the _Cylc_ GUI, which you can use to ...

```
rose suite-gcontrol &
```

!!! tip
    The `&` is optional. It detaches the invoked process, allowing the terminal prompt to remain active while the GUI is open.

> The suite itself, along will all generated output, is copied to `/scratch/<project>/${USER}/cylc-run/<suite-id>`. See the respective model documentation for details on what outputs are generated and where to find them in the output directory.

Reviewer: ```suggestion
```

Reviewer: That is true, by default the suite configuration and logs are copied there. The outputs not necessarily, but in most cases they are as well. 
And I agree, it is a piece of information that might be helpful for users to quickly know where to look for these things.

I would add it but modify it a bit. Something like:
> By default, the experiment configuration, its log files and model outputs are copied to `/scratch/<project>/${USER}/cylc-run/<suite-id>`.

> ## Editing the Suite

Reviewer: ```suggestion
## Edit the model configuration
```

To edit the suite configuration, move to the suite directory and run:

```
rose edit &
```

> This opens the _Rose_ GUI that contains setup information for the suite. The configurable options for each model are described in their respective documentation pages.

Reviewer: ```suggestion
This opens the _Rose_ GUI that contains information about the configuration. The configurable options for each model configuration are described in their respective documentation pages.
```
